ksqldb-common/src/test/java/io/confluent/ksql/function/UdfIndexTest.java

```diff
@@ -231,6 +238,127 @@ public void shouldChooseCorrectMap() {
+            SqlArgument.of(MAP1_ARG),
+            SqlArgument.of(
+                SqlLambda.of(
+                    ImmutableList.of(SqlTypes.STRING, SqlTypes.STRING),
+                    INTEGER))));
+
+    // Then:
+    assertThat(fun.name(), equalTo(EXPECTED));
+  }
+
+  @Test
+  public void shouldChooseCorrectLambdaForTypeSpecificCollections() {
+    // Given:
+    givenFunctions(
+        function(EXPECTED, false, MAP1, LAMBDA_BI_FUNCTION_STRING)
+    );
+
+    // When:
+    final KsqlScalarFunction fun1 = udfIndex.getFunction(
+        ImmutableList.of(
+            SqlArgument.of(MAP1_ARG),
+            SqlArgument.of(
+                SqlLambda.of(
+                    ImmutableList.of(SqlTypes.STRING, SqlTypes.STRING),
+                    SqlTypes.BOOLEAN))));
+
+    final KsqlScalarFunction fun2 = udfIndex.getFunction(
+        ImmutableList.of(
+            SqlArgument.of(MAP1_ARG),
+            SqlArgument.of(
+                SqlLambda.of(
+                    ImmutableList.of(SqlTypes.STRING, SqlTypes.STRING),
+                    INTEGER))));
+
+    final Exception e = assertThrows(
+        Exception.class,
+        () -> udfIndex.getFunction(
+            ImmutableList.of(
+                SqlArgument.of(MAP1_ARG),
+                SqlArgument.of(
+                    SqlLambda.of(
+                        ImmutableList.of(SqlTypes.BOOLEAN, INTEGER),
+                        INTEGER))))
+    );
+
+    // Then:
+    assertThat(fun1.name(), equalTo(EXPECTED));
+    assertThat(fun2.name(), equalTo(EXPECTED));
+    assertThat(e.getMessage(), containsString("Valid alternatives are:"
+        + lineSeparator()
+        + "expected(MAP<VARCHAR, VARCHAR>, LAMBDA<[VARCHAR, VARCHAR], A>)"));
+  }
+
+  @Test
+  public void shouldThrowOnInvalidLambdaMapping() {
+    // Given:
+    givenFunctions(
+        function(OTHER, false, GENERIC_MAP, LAMBDA_BI_FUNCTION)
+    );
+
+    // When:
+    final Exception e1 = assertThrows(
+        Exception.class,
+        () -> udfIndex.getFunction(
+            ImmutableList.of(
+                SqlArgument.of(MAP1_ARG),
+                SqlArgument.of(
+                    SqlLambda.of(
+                        ImmutableList.of(SqlTypes.BOOLEAN, SqlTypes.STRING),
+                        INTEGER))))
+    );
+
+    final Exception e2 = assertThrows(
+        Exception.class,
+        () -> udfIndex.getFunction(
+            ImmutableList.of(
+                SqlArgument.of(MAP1_ARG),
+                SqlArgument.of(
+                    SqlLambda.of(
+                        ImmutableList.of(SqlTypes.STRING,SqlTypes.STRING, SqlTypes.STRING),
+                        INTEGER)
+                )))
+    );
+
+    // Then:
+    assertThat(e1.getMessage(), containsString("Valid alternatives are:"
+        + lineSeparator()
+        + "other(MAP<A, B>, LAMBDA<[A, B], C>)"));
+    assertThat(e2.getMessage(), containsString("Number of lambda arguments doesn't match between schema and sql type"));
```

Comment:
can we give an error message here that also shows what was passed in (e.g. `LAMBDA<[STRING, STRING, STRING], INTEGER>` in this case)